### PR TITLE
add support for clan icons

### PR DESCRIPTION
**Pull request by infernalfire72**

### CODE REVIEW

File: app/api/v1/clan_icons.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+@router.get("/api/v1/clan-icons/{file_path:path}")
+@router.get("/public/api/v1/clan-icons/{file_path:path}")
+async def get_avatar(file_path: str):
+    if ".." in file_path or "/" in file_path:
+        return Response(status_code=404)
+
+    if not file_path.endswith(".png"):
+        file_path += ".png"
+
+    download_response = await s3.download(file_path, "clan-icons")
```
Comment (infernalfire72) [inline app/api/v1/clan_icons.py]: maybe smth like this?
```suggestion
    download_response = await s3.download(file_path, ImageType.CLAN_ICON.get_s3_folder())
```

File: app/api/v1/clan_icons.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import app.usecases.images
+import settings
+from app.adapters import s3
+from app.api.authorization import AdminAuthorization
+from app.api.authorization import authorize_admin
+from app.errors import Error
+from app.errors import ErrorCode
+from app.usecases.images import ImageType
+
+router = APIRouter()
```
Comment (cmyui) [inline app/api/v1/clan_icons.py:17]: needs to be included at the higher level router

File: app/api/v1/clan_icons.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        logging.warning(
+            "Unmapped error code while resolving http status code",
+            extra={"error_code": error_code},
+        )
+        return 500
+
+
+@router.post("/api/v1/clans/{clan_id}/icon")
+async def upload_avatar(
+    clan_id: str,
```
Comment (cmyui) [inline app/api/v1/clan_icons.py]: should this be an int?

File: app/api/v1/clan_icons.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    file_content: bytes = File(...),
+    authorization: AdminAuthorization = Depends(authorize_admin),
+):
+    data = await app.usecases.images.upload_image(
```
Comment (cmyui) [inline app/api/v1/clan_icons.py:41]: hmm kinda interesting but rn we don't have any checks for whether a clan exists

or whether a user exists for the avatar and profile background apis

pretty sure this is abusable, i think we should consider what to do here sometime soon if these are to be publicly exposed

File: app/api/v1/clan_icons.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+
+@router.get("/api/v1/clan-icons/{file_path:path}")
+@router.get("/public/api/v1/clan-icons/{file_path:path}")
+async def get_avatar(file_path: str):
+    if ".." in file_path or "/" in file_path:
+        return Response(status_code=404)
+
+    if not file_path.endswith(".png"):
+        file_path += ".png"
```
Comment (cmyui) [inline app/api/v1/clan_icons.py]: will it always be a png? does the upload step enforce that?

File: app/api/v1/clan_icons.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    download_response = await s3.download(file_path, "clan-icons")
+    if download_response is None:
+        logging.warning(
+            "Failed to serve non-existent clan icon",
+            extra={"file_path": file_path},
+        )
+        return Response(status_code=404)
+
+    logging.info("Served user avatar", extra={"file_path": file_path})
```
Comment (cmyui) [inline app/api/v1/clan_icons.py]: note: outdated log

File: app/api/v1/clan_icons.py (hunk 4 above)
Comment (cmyui) [reply app/api/v1/clan_icons.py:41]: for now since it's not publicly exposed we can just rely on the upstream service

**Review by cmyui — CHANGES_REQUESTED** [review]
left some comments

File: app/api/v1/clan_icons.py (hunk 3 above)
Comment (cmyui) [reply app/api/v1/clan_icons.py]: I think the avatar one should also be an int

File: app/api/v1/clan_icons.py (hunk 1 above)
Comment (cmyui) [reply app/api/v1/clan_icons.py]: maybe -- as a minimum at least use kwargs for now 
```suggestion
    download_response = await s3.download(file_path, directory="clan-icons")
```

File: app/api/v1/clan_icons.py (hunk 3 above)
Comment (infernalfire72) [reply app/api/v1/clan_icons.py]: should probably be fixed on avatars as well?

File: app/api/v1/clan_icons.py (hunk 5 above)
Comment (infernalfire72) [reply app/api/v1/clan_icons.py]: same thing for avatars

File: app/api/v1/clan_icons.py (hunk 4 above)
Comment (infernalfire72) [reply app/api/v1/clan_icons.py:41]: should probably have proper auth? lol

**Review by cmyui — APPROVED** [review]

**Pull request merged**